Quick Start booking page: Show morning and evening times separately

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* It's been reported that users sometimes get confused by AM/PM times on the booking page for Quick Start sessions. So, they sometimes end up booking up an appointment at an odd hour, like 3 AM, and then no-show. This PR is an attempt to reduce the AM/PM confusion.
* Read more in 91-gh-concierge-admin for context. Designs are also shared in that GitHub issue.

**Show filters for morning and afternoon times**

We define 12 AM - 12 PM as morning, and 12 PM - 12 AM as evening.

<img width="803" alt="Screenshot 2020-11-03 at 6 35 23 PM" src="https://user-images.githubusercontent.com/1269602/97989075-03f4a800-1e04-11eb-9431-2e8a38f457f2.png">

**Filter will be disabled if there are no open slots in that time period**

If there are no morning slots, then the morning filter will be disabled.

<img width="729" alt="Screenshot 2020-11-03 at 6 34 39 PM" src="https://user-images.githubusercontent.com/1269602/97989180-2c7ca200-1e04-11eb-967b-375e7750114b.png">

**Show optgroups in each dropdown**

This is for further making it clear early morning and morning times, as also afternoon and evening times.

<img width="331" alt="Screenshot 2020-11-03 at 6 34 56 PM" src="https://user-images.githubusercontent.com/1269602/97989313-5f269a80-1e04-11eb-8adb-f8b5270cf1e1.png">
<img width="400" alt="Screenshot 2020-11-03 at 6 35 04 PM" src="https://user-images.githubusercontent.com/1269602/97989338-66e63f00-1e04-11eb-9852-846f116e2b63.png">

**Rename 12 AM as midnight**

<img width="238" alt="Screenshot 2020-11-02 at 5 20 58 PM" src="https://user-images.githubusercontent.com/1269602/97989395-7ebdc300-1e04-11eb-9f11-3cf0c3375f0f.png">

**Some copy changes in the header**

<img width="724" alt="Screenshot 2020-11-03 at 6 34 31 PM" src="https://user-images.githubusercontent.com/1269602/97989452-8ed5a280-1e04-11eb-82d2-b01a8bb932cf.png">


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Pre-testing instructions**

Purchase a Quick Start product(likely using credits) on your wpcom account by navigating to /checkout/offer-quickstart-session. It is recommended that while testing, you are not sandboxed  so that you see the production calendar options.

**Scenario 1**
1. Visit /me/concierge to book an appointment. In the calendar options page, verify that the morning times and afternoon times filters work well, and the times are group appropriately. 12 AM to 12 PM(latter not included) is "Morning", and 12 PM - 12 AM(latter not included) is "Afternoon".
2. Ensure that the dropdown optgroup options work well. 
12 AM to 6 AM - Early morning
6 AM - 12 PM - Morning
12 PM - 6 PM - Afternoon
>  6 PM - Evening.
3. Select a few date options under Morning/Afternoon slots and verify that the appointment is booked on the correct date and time. _Note: Recommended to immediately cancel the session since the appointments are pointed to production, so HEs will get assigned to it_.

**Scenario 2**
1. Change locale to en-gb, and verify that the time options are shown in military time format i.e 00:00, 13:00 etc.
2. Repeat steps in Scenario 1 and verify they all work well here too.


Fixes 91-gh-concierge-admin

niranjan-uma-shankar committed:
WIP commit

niranjan-uma-shankar committed:
Hide the time group if no value exists

niranjan-uma-shankar committed:
Switch from segmented control to button to match the UI

niranjan-uma-shankar committed:
Switch from segmented control to button to match the UI

niranjan-uma-shankar committed:
Convert 12 AM to midnight

niranjan-uma-shankar committed:
Add title for the button when no slots available

niranjan-uma-shankar committed:
Add optgroup for each time choice

niranjan-uma-shankar committed:
Add optgroup for each time choice

niranjan-uma-shankar committed:
Add icons for the time group filter button

a8ci18n commented on client/me/concierge/book/calendar-step.js:
```diff
@@ -81,13 +82,19 @@ class CalendarStep extends Component {
+							) }
+							<br />
+							{ translate(
+								'In the meantime, consider attending one of our expert webinars on a wide variety of topics designed to help you build and grow your site. {{externalLink}}View webinars{{/externalLink}}.',
```
:warning: This change will be queued for retranslation. We'll display the English text until then.

wynwin reviewed (commented):
Looks good! Tested using a Personal Plan and didn't notice any issues. This will be a great improvement.

ivan-ottinger commented:
That's great! I didn't notice any issues in the flow. Can't wait to see this live. Thank you!

niranjan-uma-shankar committed:
A few minor fixes

sfougnier commented:
I've followed the instructions but can't seem to get past this loading screen:
<img width="761" alt="Screen Shot 2020-11-04 at 11 03 05 AM" src="https://user-images.githubusercontent.com/67021691/98134990-5201be00-1e8d-11eb-8c48-e8c0553c7a75.png">

Do you have any more specific instructions or tips for troubleshooting this? Thanks!

ivan-ottinger commented:
I run into the same issue. When I logged into my account using a different browser than usual, it worked. :)

niranjan-uma-shankar commented:
@sfougnier I believe this happens if you are using an account with 2FA enabled and using a non-wordpress, calypso.live URL. I suggest checking this PR by signing up as a new user in an incognito window. I can confirm that the loading issue will not be seen on a production wordpress.com URL.

sfougnier commented:
I finally got it to work, thanks @niranjan-uma-shankar! 

Everything looks good to me, I just have one request. Can we increase the padding between these two sentences and remove the italicized styling here?
<img width="700" alt="Screen Shot 2020-11-06 at 8 55 05 AM" src="https://user-images.githubusercontent.com/67021691/98373802-c7949800-200d-11eb-93e7-8fcc68c39bc2.png">

I think keeping the paragraphs the regular styling and adding a little more space will help make things more legible.
<img width="614" alt="Screen Shot 2020-11-06 at 8 57 32 AM" src="https://user-images.githubusercontent.com/67021691/98374038-1d694000-200e-11eb-9c20-4025052182f3.png">

niranjan-uma-shankar committed:
Styling changes

niranjan-uma-shankar commented:
Thanks, I've made the styling changes:

<img width="731" alt="Screenshot 2020-11-10 at 1 11 04 PM" src="https://user-images.githubusercontent.com/1269602/98644289-45caa480-2356-11eb-9a40-920938d0d78a.png">

sfougnier commented:
Looks good @niranjan-uma-shankar!

ddc22 commented:
Feature Testing Done !
This is definitely a great change and it works as expected. 

Small suggestion; I saw this flow for the very first time myself and I was confused as to why the time zone is mentioned in each of the options of the dropdown menu. It created an unintuitive assumption that there maybe options under different timezones. 

Maybe we can **remove the timezone from each option** and mention it once, somewhere and have ONLY the time and the AM PM (or just military time) in each of the options.
![image](https://user-images.githubusercontent.com/3422709/100101344-22404780-2e88-11eb-93b5-aac610cac9e5.png)

![image](https://user-images.githubusercontent.com/3422709/100101993-fc677280-2e88-11eb-831c-2cb792aab637.png)

I will dive into a code review next! :)

ddc22 commented on client/me/concierge/shared/available-time-card.js:
```diff
@@ -85,6 +101,80 @@ class CalendarCard extends Component {
+			options: [],
+		};
+
+		times.map( ( time ) => {
```
Semantically I think [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) is more suitable here.

ddc22 commented on client/me/concierge/shared/available-time-picker.js:
```diff
@@ -17,10 +17,28 @@ const groupAvailableTimesByDate = ( availableTimes, timezone ) => {
 		if ( dates.hasOwnProperty( startOfDay ) ) {
 			dates[ startOfDay ].times.push( beginTimestamp );
+			morningTimes.length && dates[ startOfDay ].morningTimes.push( beginTimestamp );
+			eveningTimes.length && dates[ startOfDay ].eveningTimes.push( beginTimestamp );
```
This seems a bit confusing at a glance, maybe we can use a ternary operator and add it to the relevant collection which might make it a bit more readable.
If its a concern of repetition maybe we can encapsulate the morning deciding logic inside a mini function like.

``` javascript
isMorning = (time) => {   ...   }

ddc22 commented on client/me/concierge/shared/available-time-card.js:
```diff
@@ -151,6 +253,28 @@ class CalendarCard extends Component {
+			timesForTimeGroup = morningTimes;
+			btnMorningTimeGroup = classnames( 'shared__time-group', {
+				'is-selected': isMorningTimeGroupSelected,
+			} );
```
Maybe we can just remove the use of the classnames util and hardcode the classes here since we are already inside the deciding if condition. (Same for line 273)

ddc22 commented on client/me/concierge/style.scss:
```diff
@@ -114,3 +137,16 @@
+}
+
+.calendar-step__webinars {
+margin-top: 10px;
```
Indentation broken here probably

ddc22 commented on client/me/concierge/shared/available-time-card.js:
```diff
@@ -161,29 +285,62 @@ class CalendarCard extends Component {
 					</Button>
+					<br />
+					<br />
+					{ timesForTimeGroup.length > 0 && (
```
Is this check required? It can never be false as per the current logic. Same on line 337.
If its better to keep it maybe we can move the code segment from line 338-339 inside this if condition since it is the same condition.

niranjan-uma-shankar replied:
`timesForTimeGroup` can be empty if there are no slots in the evening for example, but since the evening filter will be disabled, you are right that we can remove this check. Done in [9d35bd2](https://github.com/Automattic/wp-calypso/pull/46999/commits/9d35bd26dfde8ccbc2434b1340d1e115012169e1).

ddc22 reviewed (changes requested):
Nothing critical in code. Suggested some minor readability changes.

niranjan-uma-shankar committed:
1. Remove timezone from dropdown, and show it below the dropdown 2. Code refactoring to address review feedback

niranjan-uma-shankar commented:
Thanks for the thorough review @jdc91. I have addressed all your comments in [9d35bd2](https://github.com/Automattic/wp-calypso/pull/46999/commits/9d35bd26dfde8ccbc2434b1340d1e115012169e1).

> Maybe we can remove the timezone from each option and mention it once, somewhere and have ONLY the time and the AM PM (or just military time) in each of the options.

Good suggestion. I have removed the timezone from the dropdown options, and moved it to below the dropdown.

<img width="401" alt="Screenshot 2020-12-01 at 2 40 48 PM" src="https://user-images.githubusercontent.com/1269602/100719919-44cfe480-33e3-11eb-8d37-76f59fafc668.png">

a8ci18n commented on client/me/concierge/shared/available-time-card.js:
```diff
@@ -85,6 +102,75 @@ class CalendarCard extends Component {
+		const formattedTime = this.withTimezone( time ).format( 'LT' );
+
+		if ( [ '12:00 AM', '00:00' ].includes( formattedTime ) ) {
+			return this.props.translate( 'midnight' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **99** times:
`translate( 'Midnight' )` *ES Score: 15*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/time-indicator-concierge)

southp replied:
Is there a reason that we can't reuse "Midnight" here as the a8ci18n bot suggested?

niranjan-uma-shankar replied:
Can definitely re-use. I've pushed a change that changes this string to "Midnight".

southp commented on client/me/concierge/shared/available-time-card.js:
```diff
@@ -161,22 +283,57 @@ class CalendarCard extends Component {
+						className={ btnEveningTimeGroup }
+						onClick={ () => this.handleFilterClick( 'evening' ) }
+						disabled={ eveningTimes.length === 0 }
+						title={ btnEveningTitle }
```
Same as above :)

southp commented on client/me/concierge/shared/available-time-card.js:
```diff
@@ -161,22 +283,57 @@ class CalendarCard extends Component {
+						className={ btnMorningTimeGroup }
+						onClick={ () => this.handleFilterClick( 'morning' ) }
+						disabled={ morningTimes.length === 0 }
+						title={ btnMorningTitle }
```
I am not sure if I have seen the title attribute being used in a button tag. Could you help me understand the usage?

niranjan-uma-shankar replied:
Title attribute is allowed for <button /> tag. The reason it's used here is to provide more context when the "Morning" or "Afternoon" filter is disabled due to non-availability of slots.

southp reviewed (commented):
@niranjan-uma-shankar I can confirm that it works as expected. The code also in general LGTM. I've listed just a few more things I'd like to confirm before approving. Once you circle back, I'll follow up immediately :)

niranjan-uma-shankar committed:
Reuse string for existing translations

ddc22 commented:
I just tried en-GB and I am not seeing military time. 
![image](https://user-images.githubusercontent.com/3422709/100977525-1c70f300-3567-11eb-99e3-0d36febbf538.png)


Also in the screenshot that you have provided the time zone the timezone is IST. Shouldn't it show AM PM in those times (and not military time)?
![image](https://user-images.githubusercontent.com/3422709/100977675-4de9be80-3567-11eb-8604-4fd3d95d1de1.png)

Other than that LGTM

niranjan-uma-shankar commented:
@jdc91 The military time is based on locale, not IP. If the locale in https://wordpress.com/me/account is `English (UK)`, then military time would be shown. This locale is automatically picked up from the browser's language settings, so any user who has set `English (UK)` as their primary language in the browser would see military time, and that is why in my screenshot you see military time even though timezone is IST. There are some related discussions in the https://github.com/Automattic/wp-calypso/pull/31846. I think locale-based would cover most users, and there could be a section of users who would probably benefit from IP based. But that's probably out of scope for this PR.

southp reviewed (approved):
Thanks for answering my inquiries and addressing the string issue. No other concerns from me, as long as @jdc91 has no opinion :)

ddc22 commented:
LGTM 🚢
